### PR TITLE
Add support for Azure repositories

### PR DESCRIPTION
**Pull request by r-darwish**

### CODE REVIEW

**Comment by sshaw** [comment]
Hi, thanks for this nice addition. I'm excited to get it added.

When the repository is cloned via SSH the wrong URL is used. For example:

```
/tmp/Test >git remote -v
origin	s-shaw@vs-ssh.visualstudio.com:v3/s-shaw/Test/Test (fetch)
origin	s-shaw@vs-ssh.visualstudio.com:v3/s-shaw/Test/Test (push)
```

Results in:

https://vs-ssh.visualstudio.com/v3/s-shaw/Test/Test?path=foo.txt&version=GBmaster&line=2

Which results in a 404.

File: git-link.el [hunk 1]
```diff
@@ -166,7 +166,8 @@
     ("github" git-link-github)
     ("bitbucket" git-link-bitbucket)
     ("gitorious" git-link-gitorious)
-    ("gitlab" git-link-gitlab))
+    ("gitlab" git-link-gitlab)
+    ("visualstudio\\|azure" git-link-azure))
```
Comment (sshaw) [inline git-link.el]: Under what circumstances would one have an azure domain name?

File: git-link.el [hunk 2]
```diff
@@ -411,12 +413,30 @@ return (FILENAME . REVISION) otherwise nil."
...
 	  hostname
 	  dirname
 	  commit))
 
+(defun git-link-commit-azure (hostname dirname commit)
+  (format "https://%s/%s/commit/%s"
+	  hostname
+	  dirname
+	  (with-temp-buffer
+        (vc-git--out-ok "rev-parse" commit)
```
Comment (sshaw) [inline git-link.el]: What's the reason for calling `vc-git--out-ok` here? This is not done elsewhere in the existing code and it (seems) to work fine without it when linking to commits.

Comment (sshaw) [reply git-link.el]: Also given emacs conventions `vc-git--out-ok` is not part of the public API.

**Comment by r-darwish** [comment]
> Hi, thanks for this nice addition. I'm excited to get it added.
> 
> When the repository is cloned via SSH the wrong URL is used. For example:
> 
> ```
> /tmp/Test >git remote -v
> origin	s-shaw@vs-ssh.visualstudio.com:v3/s-shaw/Test/Test (fetch)
> origin	s-shaw@vs-ssh.visualstudio.com:v3/s-shaw/Test/Test (push)
> ```
> 
> Results in:
> 
> [vs-ssh.visualstudio.com/v3/s-shaw/Test/Test?path=foo.txt&version=GBmaster&line=2](https://vs-ssh.visualstudio.com/v3/s-shaw/Test/Test?path=foo.txt&version=GBmaster&line=2)
> 
> Which results in a 404.

Good catch. I only checked HTTP. I'll probably need to add some kind on manipulation when the link is SSH. Could you tell me that is the HTTP URL of your repository? I think that the SSH user should replace `vs-ssh`

File: git-link.el (hunk 2 above)
Comment (r-darwish) [reply git-link.el]: I use it to get the full commit hash as the Azure interface seem not to handle short commits well.

I usually run git-link from the magit log buffer. The hash that appear in that buffer is short.

Do you have another suggestion for getting the full commit hash?

File: git-link.el (hunk 1 above)
Comment (r-darwish) [reply git-link.el]: When creating a repository from Azure pipelines, such as this:

https://dev.azure.com/r-darwish/_git/testing

**Comment by r-darwish** [comment]
Thanks for replying quickly. I'll probably be able to commit the fixes in Sunday.

**Comment by sshaw** [comment]
> Could you tell me that is the HTTP URL of your repository? I think that the SSH user should replace vs-ssh

https://s-shaw.visualstudio.com/Test

File: git-link.el (hunk 1 above)
Comment (sshaw) [reply git-link.el]: Oh I see, these are more or less interchangeable:

https://dev.azure.com/s-shaw/_git/Test == https://s-shaw.visualstudio.com/_git/Test

File: git-link.el (hunk 2 above)
Comment (sshaw) [reply git-link.el]: > Do you have another suggestion for getting the full commit hash?

I will investigate.

**Comment by r-darwish** [comment]
@sshaw I pushed a commit which normalizes the SSH links. Sorry if my elisp is horrible. This is my first non-trivial contribution for an elisp package.

File: git-link.el (hunk 2 above)
Comment (sshaw) [reply git-link.el]: Thinking back and looking though Emacs' docs –and this code– I recall that at one point I thought: "I should have just used Emacs's `vc-xxx` [or whatever] functions to interact with git" but then I found that there isn't a well-defined public API for this type of usage.   

The way I have done things like this is to make a function to call `git`. There are [several examples of this](https://github.com/sshaw/git-link/blob/5581b4760aa2efb01eddca035a2758bc62da8635/git-link.el#L209) in the code now.

File: git-link.el [hunk 3]
```diff
@@ -411,12 +428,30 @@ return (FILENAME . REVISION) otherwise nil."
...
+        (string-trim (buffer-string)))))
```
Comment (sshaw) [inline git-link.el]: Why is the call to `buffer-string` needed? The functions should depend on their args and not on the caller's context.

**Comment by r-darwish** [comment]
Done 🙂

**Comment by sshaw** [comment]
Great, thanks so much!

**Comment by sshaw** [comment]
Released in [v0.7.5](https://github.com/sshaw/git-link/releases/tag/v0.7.5)

**Pull request merged**